### PR TITLE
Pro tiers and liq fee support

### PR DESCRIPTION
**Pull request by gdev8317**
Changelog:

- added support for liquidation fees configurable per market. it is configured as % of position size charged on liquidation
- pro tiers:
  - if a trader has pro tier they have discount for open/close fee
  - affiliate reward is capped at max of (pro discount, total referral rebate)
  - there is configurable min affiliate reward per referral tier level. for example, if it is set to 5%, then affiliate is guaranteed to receive to receive 5% of open/close fee even if pro discount exceeds total referral rebate
- added `StopIncrease` order. works similar to `LimitIncrease` but trigger when price moves in the opposite direction
- added `validFromTime` param, it allows to support for TWAP orders
- split configuration for swap, discount and withdrawal fee factor, they are now configured separately
- supported enabling/disabling open interest / max open interest in utilisation calculation
- fixed `MAX_AUTO_CANCEL_ORDERS` validation that caused failed order updates if number of auto cancel orders was at max
- removed separate execution fee for single / double token deposit
- removed block numbers from entities to save gas
- removed unused `Config.setFundingRate`
- removed unused `Keys.MAX_FUNDING_FACTOR_PER_SECOND_LIMIT`
- added `MAX_FUNDING_FACTOR_PER_SECOND` to `allowedLimitedBaseKeys`
- added nonReentrant modifier for `ExchangeRouter.makeExternalCalls`
- added `autoCancel` flag to `emitOrderCreated`, `emitOrderUpdated`, removed emitting of `isFrozen` flag from `emitOrderCreated`
- added simulateExecute*Latest functions that don't require `key` 

Important for integrations / UI:
- structure of `Order`, `Deposit` and `Withdrawal` has changed and signature of callback functions has changed as well
- the `estimateExecuteDepositGasLimit` function has been updated so that single sided deposits use the same gas expenditure as double sided gas deposits.
- additionally the `depositGasLimitKey` has been changed to no longer accept a boolean indicating whether it is a single sided deposit.
- `simulateExecute*Latest` functions can be used instead of `simulateExecute*` to speed up simulation, there is no need to request nonce and pre-calculate new order key 
- format of some events was changed. `PositionFeesInfo` and `PositionFeesCollected` have dynamic set of fields. Some existing fields were renamed

### CODE REVIEW

File: contracts/config/Config.sol [hunk 1]
```diff
@@ -679,9 +676,11 @@ contract Config is ReentrancyGuard, RoleModule, BasicMulticall {
             baseKey == Keys.POSITION_FEE_RECEIVER_FACTOR ||
             baseKey == Keys.SWAP_FEE_RECEIVER_FACTOR ||
             baseKey == Keys.BORROWING_FEE_RECEIVER_FACTOR ||
+            baseKey == Keys.LIQUIDATION_FEE_RECEIVER_FACTOR ||
```
Comment (xvi10) [inline contracts/config/Config.sol:695]: seems not needed since LIQUIDATION_FEE_FACTOR is validated above

File: contracts/data/Keys.sol [hunk 2]
```diff
@@ -375,6 +381,7 @@ library Keys {
     bytes32 public constant CLAIMABLE_COLLATERAL_TIME_DIVISOR = keccak256(abi.encode("CLAIMABLE_COLLATERAL_TIME_DIVISOR"));
     // @dev key for claimed collateral amount
     bytes32 public constant CLAIMED_COLLATERAL_AMOUNT = keccak256(abi.encode("CLAIMED_COLLATERAL_AMOUNT"));
+    bytes32 public constant USAGE_FACTOR_IGNORE_OPEN_INTEREST = keccak256(abi.encode("USAGE_FACTOR_IGNORE_OPEN_INTEREST"));
```
Comment (xvi10) [inline contracts/data/Keys.sol]: suggestion for an alternative name: IGNORE_OPEN_INTEREST_FOR_USAGE_FACTOR

File: contracts/position/DecreasePositionCollateralUtils.sol [hunk 3]
```diff
@@ -610,67 +611,13 @@ library DecreasePositionCollateralUtils {
...
-            0, // protocolFeeAmount
-            0, // positionFeeReceiverFactor
-            0, // feeReceiverAmount
-            0, // feeAmountForPool
-            0, // positionFeeAmountForPool
-            0, // positionFeeAmount
-            0, // totalCostAmountExcludingFunding
-            0 // totalCostAmount
-        );
+        PositionPricingUtils.PositionFees memory _fees;
```
Comment (xvi10) [inline contracts/position/DecreasePositionCollateralUtils.sol]: i don't think this should be changed, as we need the `fees.funding.claimableLongTokenAmount`, `fees.funding.claimableShortTokenAmount`, etc?

File: contracts/config/Config.sol (hunk 1 above)
Comment (gdev8317) [reply contracts/config/Config.sol:695]: i think it still makes sense to validate both LIQUIDATION_FEE_RECEIVER_FACTOR and LIQUIDATION_FEE_FACTOR?

File: contracts/position/DecreasePositionCollateralUtils.sol (hunk 3 above)
Comment (gdev8317) [reply contracts/position/DecreasePositionCollateralUtils.sol]: fixed

**Pull request merged**